### PR TITLE
Change API to allow multiple simultaneous clients

### PR DESCRIPTION
**Pull request by tomaka**
Right now the way the `Transport` trait works is that it produces a stream of sockets over which modifiers are applied in order to negotiate a protocol. Unfortunately an overlook of this design is that elements of a stream are polled serially, and not in parallel. The consequence is that right now a server cannot accept any further client while a client is already being negotiated.

This PR fixes that by changing the item yielded by the `Listener` stream from a `Result<impl AsyncRead + AsyncWrite, IoError>` to a `impl Future<Item = impl AsyncRead + AsyncWrite, Error = IoError>`.

Two notes:
- This doesn't change the various `dial_and_listen` methods because a general improvement of multiplexing is needed anyway.
- The changes to `ConnectionReuse` introduced by this PR are very unsatisfactory, for the same reason.

### CODE REVIEW

**Comment by tomaka** [comment]
For the record this doesn't allow the echo example to handle multiple simultaneously clients, because the code to do so is non-trivial (requires creating a custom implementation of `Stream` or `Future`).

File: example/examples/echo-server.rs [hunk 1]
```diff
@@ -95,62 +95,49 @@ fn main() {
...
+
+                    // We loop forever in order to handle all the messages sent by the client.
+                    loop_fn(socket, move |socket| {
+                        let client_addr = client_addr.clone();
+                        socket.into_future()
+                            .map_err(|(err, _)| err)
+                            .and_then(move |(msg, rest)| {
+                                if let Some(msg) = msg {
+                                    // One message has been received. We send it back to the client.
+                                    println!("Received a message from {}: {:?}\n => Sending back \
```
Comment (folsen) [inline example/examples/echo-server.rs]: Is this just for debugging now or should we bring in a logging library like the parity client, allowing users to flip some flags for extra debug data? Just a general question that I thought of when I saw this, doesn't have to be addressed in this PR obviously.

Comment (tomaka) [reply example/examples/echo-server.rs]: This example is here to quickly demonstrate how to use libp2p, so I don't think a logging library is necessary.

Comment (folsen) [reply example/examples/echo-server.rs]: No definitely not, I was just talking generally, might be nice to be able to flip on something like trace messages, we can move that conversation to Riot :P

**Pull request merged**